docs/markdown/Ad-Hoc Tools/adhoc-tool.md

```diff
@@ -0,0 +1,98 @@
+
+To get the best cache efficiency, it can make sense to break your `adhoc_tool` into smaller incremental steps. For example, if your process needs to fetch dependencies and then build a library based on those dependencies and some first-party source files, having one `adhoc_tool` for each of those steps means that the dependency-fetching stage will only be re-run when your requirements change, and not when the first-party source files change.
+
+Generally, if you are chaining `adhoc_tool` targets together , it will be easier to use the default `workdir` and `root_output_directory` fields for each step that will be consumed by an `adhoc_tool` in the same `BUILD` file. Change the `root_output_directory` only for targets that are intended to be used in other places or ways.
+
+
+## Wrapping generated sources for use by other targets
+
+_Our [Antlr demo](https://github.com/pantsbuild/example-adhoc/tree/main/antlr) demonstrates wrapping the outputs of `adhoc_tool` targets for use as Python sources._
+
+`adhoc_tool` generates `file` sources by default. This can be acceptable if generating assets that do not need to be consumed as source files for another Pants backend. Other Pants backends need generated sources to be marked as actual source files.
+
+There are several targets included in Pants with the prefix `experimental_wrap_as_`. These act as a source target that can be used as a dependency in a given language backend, with the caveat that dependency inference is not available.
+
+
+## Using externally-managed tools
+
+_(Our [JavaScript demo](https://github.org/pantsbuild/example-adhoc/tree/main/javascript)) demonstrates the use of externally-managed binaries._
+
+Some build processes need to make use of tools that can't be modeled within a Pants codebase. The `system_binary` target lets you make use of a binary that is installed on the system. `system_binary` targets may be specified as `runnable` or `runnable_dependency` values for `adhoc_tool`.
+
+`system_binary` will search for a binary in pre-defined or user-supplied search paths with a given `binary_name`. To improve reproducibility, it's possible to test matching binaries with sample arguments, to see if its output matches a given regular expression. This can be used to match against version strings. If a such a binary does not exist on the system where Pants is being run, any build involving this `system_binary` target will fail.
+
+When specified as a `runnable_dependency`, the binary will be available on the `PATH` with the target name of the dependency. This can be important if the `runnable` field invokes a subprocess (for example, `yarn` tries to invoke a binary called `node` as its interpreter).
+
+
+## Running shell scripts
+
+Currently, `shell_source` targets are not runnable. In the meantime, it is possible to run a shell script as an `adhoc_tool` through the following approach:
```

Comment:
This is handy, what do you think of having something similar in `docs/markdown/Shell/run-shell-commands.md` to point people to the better `adhoc_tool` instead of `shell_command` for running in-repo processes? I imagine might try running a tool like antlr by loading it up into a shell command, since shell is the usual way people run things like that.

(Also, not directly related to this PR, but while looking there, I happened to notice that that page still has the `🚧 Feedback wanted` referring to experimental status, which probably doesn't apply any more.)